docs/painless/painless-guide/painless-datetime.asciidoc

```diff
@@ -0,0 +1,295 @@
+[[painless-datetime]]
+=== Using Datetime in Painless
+
+==== Datetime Representation
+
+Datetimes in Painless are most commonly represented as a
+<<primitive-types, long>>, a <<string-type, String>>, or a
+<<painless-api-reference-shared-ZonedDateTime, ZonedDateTime>>.
+
+long:: represents a datetime as the number of milliseconds or nanoseconds since
+epoch (1970-01-01T00:00:00Z)
+String:: represents a datetime as a sequence of characters defined by a
+well-known standard such as https://en.wikipedia.org/wiki/ISO_8601[ISO 8601] or
+defined by the source of input in a custom way
+ZonedDateTime:: a <<reference-types, reference type>> (object) that contains an
+internal representation of a datetime and provides numerous
+<<painless-api-reference-shared-ZonedDateTime, methods>> for
+modification and comparison
+
+A typical pattern in a script is to convert a long or String representation of
+a datetime to a ZonedDateTime representation, modify, or compare the
+ZonedDateTime representation, and then format it back to a long or String
+representation for storage or as a returned result depending on the script's
+objective(s).
+
+==== Datetime API
+
+Datetimes in Painless use the standard Java libraries and are available through
+the Painless <<painless-api-reference-shared, Shared API>>. Most of the classes
+from the following Java packages are available to use in Painless scripts:
+
+* <<painless-api-reference-shared-java-time, java.time>>
+* <<painless-api-reference-shared-java-time-chrono, java.time.chrono>>
+* <<painless-api-reference-shared-java-time-format, java.time.format>>
+* <<painless-api-reference-shared-java-time-temporal, java.time.temporal>>
+* <<painless-api-reference-shared-java-time-zone, java.time.zone>>
+
+==== Datetime Parsing and Formatting
+
+Switching between different representations of datetimes is often necessary to
+achieve a script's objective(s). Datetime parsing is a switch from a long
+or String to a ZonedDateTime, and datetime formatting is a switch from a
+ZonedDateTime to a long or a String.
+
+A <<painless-api-reference-shared-DateTimeFormatter, DateTimeFormatter>> is a
+<<reference-types, reference type>> (object) that defines the allowed sequence
+of characters for a String representation of a datetime. Switching of datetime
+representations between a String and a ZonedDateTime and vice versa often
+requires a DateTimeFormatter. For more information about how to use a
+DateTimeFormatter see the
+{java11-javadoc}/java.base/java/time/format/DateTimeFormatter.html[Java documentation].
+
+===== Datetime Parsing Examples
+
+* parse from a long (milliseconds)
```

Comment:
Maybe also have a "parse from milliseconds in a string" or something like that

Reply:
Added.